### ISSUE TITLE
Roadmap and new features

### ISSUE THREAD

**Issue by franciscop** [umbrellajs/umbrella#17]
| issue       | done | merged | description   |
|--------------|------|--------|---------------|
| [x]            | [x]    | [x]    | better `ajax()` and `.ajax()` |
| [x]            | [x]    | [x]    | better `serialize()` |
| [x]            | [x]    | [x]    | `.scroll()` to scroll to a part of the page, smoothly if supported |
| [x]            | [x]    | [x]    | `.size()` to later get the `width`, `height`, `top`, `left`, `right`, `bottom` |
| [x]            | [x]    | [x]     | `.off()` to remove events |
| [x]            | [x]    | [x]     | `.text()` to get or set the html text (similar to `.html()`) |
| [x]            | [x]    | [x]     | pass data to `.trigger()` as second parameter, migrate to CustomEvent |
| [x]            | [x]    | [x]     | HTML generator as `u('<div class="test"></div>')` creates the specified nodes. It is not really useful as nothing accepts them, but it does create them. |

Ongoing (always need some work):

- **documentation**
- **tests**
- **examples**, preferably real-world ones

***

Original comment:

So, what features do you think Umbrella JS is missing? Take into account that **it's not a jquery clone**, it just reuses the most convenient method names. While recent `.not()` is perfectly valid, `.siblings()` is in the limit of the scope of the library and things like `.prev()` are definitely outside.

So, what's the scope? I am not sure how to define it properly, but here I'll try:

- Features that are broadly applicable to javascript world (including the "most used" subset of jquery)
- Small and reusable features. Anything that deletes code is welcome
- Features that are trivial to implement by the user are excluded. This includes things like `get(n)`, for example, since you can just do `.nodes[n]`.

I think with these I'd be happy and I'm planning to do them in the future. Any other suggestion? @tr3intaydos

**Comment by tr3intaydos** [comment]
For DOM selectors all it really needs is up, down and sideways, and its done.

1. .show() and .hide() (may be needed some sort of .css('display')
2. There is .on() but not .off()
3. scroll event and position?
4. .clone()
5. .width() and .height()
6. .ready()
7. .text()
8. HTML Generator
9. Documentation and the Tests could be improved.
10. Maybe a 'custom umbrella.min.js' generator? Only the plugin you need and dependencies
11. Some real-life applications
12. Type detection? Like [$.type(obj);](https://api.jquery.com/jQuery.type/)

**Comment by franciscop** [comment]
1. [no] Style functions are outside the scope of Umbrella. Modern CSS3 is very capable of doing most of the jquery style work. Instead of `.show()`, an `.addClass('visible')` should be considered, for example. Or `.removeClass('hidden')`. Doing styles in JS is quite bad practice from a Separation of Concerns standpoint.
1. [yes] `.off()` [should be easy to implement](http://youmightnotneedjquery.com/#off) and it's within the scope
1. [yes] I just learned yesterday about `scrollIntoView()`, which is trivial to implement: `u(to).first().scrollIntoView({behavior: "smooth"});`. [Support for the smooth parameter](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView#Browser_compatibility) is awful so we cannot support it by default, but otherwise is retro-compatible up to IE6. If you want to add it go ahead (:
1. [maybe] .clone(), as well as an HTML Generator, is useful mainly with `appendTo` and similar methods. This is a whole section that I'm not sure if it's whithin the scope, but it might be. I need to think about this part. See [2]
1. [changed] `.width()` and `height()`, as well as `top`, `left`, `right` and `bottom` can be obtained straight from [`getBoundingClientRect()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/getBoundingClientRect), which is great. I think they all could be abstracted to a single method instead of having to specify one for each of them, like `.rect()` or `.size()`
1. [no] There are [so](https://css-tricks.com/snippets/javascript/cross-browser-dependency-free-dom-ready/) [many](https://github.com/franciscop/pagex) [libraries](https://dojotoolkit.org/reference-guide/1.10/dojo/ready.html) for `ready()` that I think it's best if it stays out. Also [it's not needed](http://stackoverflow.com/a/6026730) if you put your scripts at the end of the page
1. [yes] `.text()` could be implemented with `.textContent`, however I'm trying few things that might make it not so neccessary [1]
1. [maybe] See 4.
1. [yes] Completely agree, I'm trying to do that but it's a huge job
1. [no] That's really good concept, but I think it's to solve a problem that we don't have (and I expect not to have), which is library size. [2]
1. [yes] There are many things to document, including real-life. I think that few of the denied features could be a small tutorial for real-life examples instead.
1. [no] I was thinking of type detection for internal use, but not for external one. It's quite different, see the source of `.slice()` in `umbrella.js` for example. What something might be enough "array-like" for one situation, it might not be for another. I think this is better for this to stay out of the public API since there are too many "one-off" decisions to make here.

Legend:

- [yes] Go ahead if you want, please create an issue first as I'm likely to also try some of them
- [change] Yes but with small changes (they are explained)
- [maybe] Let me think about it, better don't do it yet
- [no] This is outside the scope of Umbrella (or other, see explanation)


[1] There's [something that can be made](https://github.com/umbrellajs/umbrella/issues/10#issuecomment-174782471) that makes it really easy to perform this in another way:

```js
// Get the first child's html or set the html of all of them
var html = u('a').innerHTML;
u('a').innerHTML = 'whatever';

// Get the first child's text or set the text of all of them
var text = u('a').textContent;
u('a').textContent = 'whatever';
```

I am thinking whether or not it'd be a good idea to make this for default elements such as `.html` or `.text`:

```js
var html = u('a').html;
u('a').html = 'whatever';
```

However it has the disadvantage that the `set` way is not chainable, which is one of the main advantages of this type of libraries.



[2] I think in JS land things are changing to either use a *huge* library (Angular, React, jQuery+plugins?, etc) or to just mix and match many small libraries you need. This might (or not) be influenced by Node since that's what is becoming the common practice there. I love this and even now I think that Umbrella is doing too many things, since it's doing four things:
- DOM transversal
- DOM editing (not manipulation)
- Events
- Ajax

As I'm thinking of taking Ajax out of the equation, I am not sure how I feel about don't think adding DOM manipulation there. Also if it stays within 2-3kb of size, I think that custom builds would be more pain than what is worth it long-term (I tried that in [Picnic CSS](http://picnicss.com/)).

**Comment by tr3intaydos** [comment]
I got something working with .clone() but the trouble is at using .append()/.after()/.before(), they should accept objects, not only html

**Comment by franciscop** [comment]
I think this is really interesting for that, I didn't realize of it until I saw it in jquery:

```text
htmlString or Element or Array or jQuery
```

There is one element that already accepts all of these methods (except last one, which can be easily implemented) which is the constructor `u()`. Maybe all of the parts of the code that accept right now a selector such as `.whatever('#hello .world')` could be reimplemented to using `u()`, including `.append()`, `.after()`, etc

**Comment by CraigDMcKenna** [comment]
Am I mistaken or is [#34](https://github.com/umbrellajs/umbrella/issues/34) complete now?

**Comment by franciscop** [comment]
@CraigDMcKenna you are right, it's finished (in the branch "2"). 

About the documentation, I just reviewed it and I'm surprised that the documentation seems more completed than expected. Only the method `.serialize()` has no documentation at all.

However, there are MANY tests missing. Basically, every folder within `src/core` (old `src/plugins`) should have a test. Some would be a higher priority than others, for instance **tests for those are much needed**:

- `param()`: Parametize an object: `{ a: 'b', c: 'd' }` => `'a=b&c=d'`
- `unique()`: Removed duplicated nodes, used for some specific methods
- `.slice()`: Force a parameter to become an array AND also it clones them

The `.slice()` method [will be changed soon](https://github.com/umbrellajs/umbrella/issues/44), but it should keep a really similar footprint so most/all tests will be valid in the new implementation.
